lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -1184,11 +1184,10 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
       return OpaqueType;
     }
 
-    auto *opaqueType = createOpaqueStructWithSizedContainer(
-        Scope, Decl ? Decl->getNameStr() : "", File, Line, SizeInBits,
-        AlignInBits, Flags, MangledName, collectGenericParams(Type),
-        UnsubstitutedType);
-    return opaqueType;
+    auto *OpaqueType = createOpaqueStruct(
```

Comment:
It looks like this code doesn't look into the parent type of the nominal type at all, so for example if you have

```swift
struct Outer<T> {
  struct InnerGeneric<U> {}
  struct Inner {}
}
```

The type `Outer<Int>.Inner` is a StructType and not a BoundGenericStructType, but it _is_ specialized (isSpecialized() returns true) and getParent() returns the BoundGenericStructType `Outer<Int>`.

Reply:
In this case "opaque type" is in regards to the debug info representation of the type.  An opaque entry is one that doesn't contain any information about the type's fields, only some basic information about the type (like name, size, alignment, etc). The compiler emits complete debug info for the unsubstituted generic type, and then an opaque entry for each substituted type.

Reply:
In that case why does the substituted type need to record generic arguments at all? Either you need the complete set and you should walk into the parent type, or you don't need them at all. Just collecting the innermost arguments doesn't seem to make much sense.

Reply:
Today we collect them to ensure they survive in the debug info when producing a dSYM. dsymutil may delete any debug info that is unreachable when stitching together debug info from multiple .o files, so if a type happens to only be used as a generic argument, it would be deleted from the debug info.

In the future we might also stop emitting the substituted type's mangled name, and produce it by combining the unsubstituted generic type with the generic arguments.

Reply:
> Today we collect them to ensure they survive in the debug info when producing a dSYM. 

It seems that by the same reasoning you should collect the parent type’s generic arguments too then.

Reply:
Oh I see what you mean now. You're right, this doesn't address an inner specialized non-generic type at all. I'll address that in a follow up patch.